src/coreclr/nativeaot/Runtime/eventtrace_bulktype.cpp

```diff
@@ -69,64 +68,53 @@ void BulkTypeEventLogger::FireBulkTypeEvent()
     UINT16 nClrInstanceID = GetClrInstanceId();
 
-    UINT iDesc = 0;
-
-    _ASSERTE(iDesc < _countof(EventData));
-    EventDataDescCreate(&EventData[iDesc++], &m_nBulkTypeValueCount, sizeof(m_nBulkTypeValueCount));
+    if(m_pBulkTypeEventBuffer == NULL)
+    {
+        // The buffer could not be allocated when this object was created, so bail.
+        return;
+    }
 
-    _ASSERTE(iDesc < _countof(EventData));
-    EventDataDescCreate(&EventData[iDesc++], &nClrInstanceID, sizeof(nClrInstanceID));
+    UINT iSize = 0;
 
     for (int iTypeData = 0; iTypeData < m_nBulkTypeValueCount; iTypeData++)
     {
+        BulkTypeValue& target = m_rgBulkTypeValues[iTypeData];
+
         // Do fixed-size data as one bulk copy
-        _ASSERTE(iDesc < _countof(EventData));
-        EventDataDescCreate(
-            &EventData[iDesc++],
-            &(m_rgBulkTypeValues[iTypeData].fixedSizedData),
-            sizeof(m_rgBulkTypeValues[iTypeData].fixedSizedData));
+        memcpy(
+                m_pBulkTypeEventBuffer + iSize,
+                &(target.fixedSizedData),
+                sizeof(target.fixedSizedData));
+        iSize += sizeof(target.fixedSizedData);
 
         // Do var-sized data individually per field
-
-        // Type name (nonexistent and thus empty on nativeaot)
-        _ASSERTE(iDesc < _countof(EventData));
-        EventDataDescCreate(&EventData[iDesc++], L"", sizeof(WCHAR));
+        // No name in event, so just the null terminator
+        LPCWSTR wszName = NULL;
+        m_pBulkTypeEventBuffer[iSize++] = 0;
+        m_pBulkTypeEventBuffer[iSize++] = 0;
 
         // Type parameter count
-        _ASSERTE(iDesc < _countof(EventData));
-        EventDataDescCreate(
-            &EventData[iDesc++],
-            &(m_rgBulkTypeValues[iTypeData].cTypeParameters),
-            sizeof(m_rgBulkTypeValues[iTypeData].cTypeParameters));
+        DWORD cTypeParams = target.cTypeParameters;
+        ULONG *ptrInt = (ULONG*)(m_pBulkTypeEventBuffer + iSize);
+        *ptrInt = cTypeParams;
+        iSize += 4;
 
         // Type parameter array
-        if (m_rgBulkTypeValues[iTypeData].cTypeParameters > 0)
+        if (cTypeParams == 1)
+        {
+            memcpy(m_pBulkTypeEventBuffer + iSize, &target.ullSingleTypeParameter, sizeof(ULONGLONG) * cTypeParams);
+            iSize += sizeof(ULONGLONG) * cTypeParams;
+        }
+        else
         {
-            _ASSERTE(iDesc < _countof(EventData));
-            EventDataDescCreate(
-                &EventData[iDesc++],
-                ((m_rgBulkTypeValues[iTypeData].cTypeParameters == 1) ?
-                    &(m_rgBulkTypeValues[iTypeData].ullSingleTypeParameter) :
-                    (ULONGLONG*)(m_rgBulkTypeValues[iTypeData].rgTypeParameters)),
-                sizeof(ULONGLONG) * m_rgBulkTypeValues[iTypeData].cTypeParameters);
+            memcpy(m_pBulkTypeEventBuffer + iSize, target.rgTypeParameters.GetValue(), sizeof(ULONGLONG) * cTypeParams);
```

Comment:
Where does `rgTypeParameters` actually get updated? I'm missing where anything actually appends to it.